base/strings/string.jl

```diff
@@ -284,23 +284,25 @@ getindex(s::String, r::AbstractUnitRange{<:Integer}) = s[Int(first(r)):Int(last(
     return ss
 end
 
-length(s::String) = length_continued(s, 1, ncodeunits(s), ncodeunits(s))
+# nothrow because we know the start and end indices are valid
+@assume_effects :nothrow length(s::String) = length_continued(s, 1, ncodeunits(s), ncodeunits(s))
 
-@inline function length(s::String, i::Int, j::Int)
+# effects needed because @inbounds
+@assume_effects :consistent :effect_free @inline function length(s::String, i::Int, j::Int)
```

Comment:
Why is this function `:consistent` even if it applies `@inbounds` on `thisind`, which contains `@boundscheck`?

Reply:
because it's does so only after a bounds check to ensure that the indexing is inbounds.

Reply:
I don't quite follow that explanation yet. Shouldn't the presence of `@boundscheck` still taint the caller, if the caller used `@inbounds`? Or is that the `noinbounds` effect bit (currently that is not mentioned in the docstring for `@assume_effects`)